### PR TITLE
Add war generator function

### LINKED ISSUE #14: Implement generation of WAR using CWP libraries
The CWP libraries make it much easier to do this step, though it takes a lot of time. Something that we need to think about.

### PR DESCRIPTION
**Pull request by sladyn98**
This PR adds the war generator function using the custom war packager library. It pulls in the packager config from the resources folder.

The Jenkinsfile Needs to be fixed before this can be merged in
Closes #14

### CODE REVIEW

**Comment by sladyn98** [comment]
> I can see this working. Eventually there probably should be a way to clean out /tmp

@kwhetstone Should we do it after the user has downloaded it, But one minor inconvenience that would cause is that if the user hits download multiple times then we wouldn't have it in temp and would have to regenerate it

**Comment by sladyn98** [comment]
@kwhetstone I guess we can merge this once the build passes :)

File: src/test/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGeneratorTest.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.org.jenkins.custom.jenkins.distribution.service.generators;
+
+import org.junit.jupiter.api.Test;
+
+public class WarGeneratorTest {
+
+    @Test
+    public void testWarGenerationSucceeds() {
+        WarGenerator.generateWAR("1.0");
```
Comment (sladyn98) [inline src/test/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGeneratorTest.java:9]: @kwhetstone @LinuxSuRen When I try to test this the builder throws a huge error 
```
java.io.IOException: Command failed with exit code 1: mvn clean package
	at io.jenkins.tools.warpackager.lib.util.SystemCommandHelper.processFor(SystemCommandHelper.java:39)
	at io.jenkins.tools.warpackager.lib.util.MavenHelper.run(MavenHelper.java:66)
	at io.jenkins.tools.warpackager.lib.util.MavenHelper.run(MavenHelper.java:46)
	at io.jenkins.tools.warpackager.lib.impl.Builder.build(Builder.java:146)
	at com.org.jenkins.custom.jenkins.distribution.service.generators.WarGenerator.generateWAR(WarGenerator.java:24)
	at com.org.jenkins.custom.jenkins.distribution.service.generators.WarGeneratorTest.testWarGenerationSucceeds(WarGeneratorTest.java:9)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
```

File: src/main/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGenerator.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    private static final String DEFAULT_TMP_DIR_NAME = "tmp";
+    private static final String PACKAGER_CONFIG_YAML = "packager-config.yml";
+
+    public static void generateWAR(String versionName) {
+        final Config cfg;
+        final File configPath = util.getFileFromResources(PACKAGER_CONFIG_YAML);
+        try {
+            cfg = Config.loadConfig(configPath);
+            cfg.buildSettings.setTmpDir(new File(DEFAULT_TMP_DIR_NAME));
```
Comment (LinuxSuRen) [inline src/main/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGenerator.java:21]: You can get a temporary directory by the following code:
`System.getProperty("java.io.tmpdir")`
Furthermore, a user can give it a different value by passing a path. For example: `java -Djava.io.tmpdir=/path/to/tmpdir`

Comment (sladyn98) [reply src/main/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGenerator.java:21]: Yes made the change :+1:

**Comment by sladyn98** [comment]
@LinuxSuRen The test is throwing errors.

**Comment by LinuxSuRen** [comment]
It might be caused by some restriction OS environments.

**Comment by sladyn98** [comment]
Maybe that is why it asks for a `tmp` directory within the project structure and not the system temp directory, should we just revert to using the project temp directory instead of the system one ?

**Comment by LinuxSuRen** [comment]
> Maybe that is why it asks for a `tmp` directory within the project structure and not the system temp directory, should we just revert to using the project temp directory instead of the system one ?

I think that you can pass it another directory.

**Comment by sladyn98** [comment]
> I think that you can pass it another directory. It is already configrable now.

Like what ? Because it would anyway be created in the project structure, so does it matter if I pass temp again ?

**Comment by LinuxSuRen** [comment]
Or you can get the temporary directory by reading a config file.

**Comment by sladyn98** [comment]
> Or you can get the temporary directory by reading a config file.

I do not think it would make a difference since the value in the config file would be static

**Comment by kwhetstone** [comment]
If using Java you can write a temp file using [Files](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html).  I think all you have to do here is call delete on the file object at the end of all this.

**Comment by kwhetstone** [comment]
@sladyn98 let's update the Jenkinsfile and see if that fixes the problem.
[lines](https://github.com/jenkins-infra/plugin-site-api/blob/master/Jenkinsfile#L47-L54)

File: src/test/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGeneratorTest.java (hunk 1 above)
Comment (sladyn98) [reply src/test/java/com/org/jenkins/custom/jenkins/distribution/service/generators/WarGeneratorTest.java:9]: Open issue in Custom War Packager.

**Comment by kwhetstone** [comment]
@sladyn98 Looks like there's an error while trying to build the job. `org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [com.org.jenkins.custom.jenkins.distribution.service.CustomJenkinsDistributionServiceApplication]; nested exception is org.springframework.context.annotation.ConflictingBeanDefinitionException: Annotation-specified bean name 'updateCenterService' for bean class [com.org.jenkins.custom.jenkins.distribution.service.services.UpdateCenterService] conflicts with existing, non-compatible bean definition of same name and class [com.org.jenkins.custom.jenkins.distribution.service.UpdateCenterService]`

The reason the build failed was looking for test reports and it couldn't find it.  You can configure `archiveArtifacts allowEmptyArchive: true, artifacts: '**/target/**/*.jar`

**Pull request merged**